### ISSUE TITLE
macOS: Incorrectly scales the screen on a non-retina display

### ISSUE THREAD

**Issue by danbee** [joncampbell123/dosbox-x#1697]
**Describe the bug**
DOSBox-X running on macOS seems to double the size of the display on a non-retina screen attached to a retina mac. This means you can only see the bottom left corner of whatever is running.

**To Reproduce**
1. Connect a non-retina screen to a retina mac (I'm using a Thunderbolt Display)
2. Load DOSBox-X
3. If the DOSBox-X window opens on the laptop's display, move the window to the external non-retina display and resize it to trigger a redraw

**Expected behavior**
I expect to see the DOS screen correctly scaled for the window. Moving the window onto the laptop's display and resizing it causes the screen to be scaled correctly.

**Screenshots**

DOSBox-X on the MacBook's built in retina display:
![image](https://user-images.githubusercontent.com/165531/85959828-c7540400-b96c-11ea-9ae3-4f502c6aa326.png)

DOSBox-X on the external non-retina Thunderbolt Display:
![image](https://user-images.githubusercontent.com/165531/85959741-2cf3c080-b96c-11ea-9315-113da6d85255.png)

**Environment (please complete the following information):**
 - macOS 10.15 Catalina running on a 15", 2017 MacBook Pro
 - DOSBox-X 0.83.2 installed from Homebrew Cask
 - Default configuration

**Additional context**

**Comment by joncampbell123** [comment]
Does output=surface work?

Mac OS X's OpenGL support vs retina/non-retina is a pain to support because it seems wildly inconsistent across major releases how to compensate for it.

**Comment by danbee** [comment]
Hi @joncampbell123, yes, output=surface works just fine, although it doesn't scale the image up to fill the screen. It doesn't always render the image in the centre though, especially if I select the "Force scaler" option.

I'm going to take a look and see if I can figure out how to fix it (although my SDL/OpenGL fu is weak!). Let me know if you want me to try any debug builds.

**Comment by danbee** [comment]
Hmmm, if I build the project with SDL2 instead the screen size is correct on both displays. The app is pretty unstable though! I guess this points to a difference in the way SDL1 and SDL2 report screen metrics?

**Comment by joncampbell123** [comment]
Whatever SDL2 is doing right, needs to be backported to the highly modified SDL1 in DOSBox-X.

**Comment by eirnym** [comment]
Also SDL2 incorrectly scales on Retina display, like this, fullscreen is even worse.

<img width="832" alt="Zrzut ekranu 2020-07-21 o 00 15 43" src="https://user-images.githubusercontent.com/485399/87991904-54f7bf00-cae7-11ea-9716-86bffe31aac1.png">

**Comment by john-at-charpa** [comment]
There are some projects that exist which may be of some assistance here. This project has SDL2 implemented and has both metal rendering (macOS) as well as retina display fixes in the code:

[https://github.com/duganchen/dosbox](url)

For example, from sdlmain.cpp:
```
#if defined(C_RETINAFIX)
      sdl.window = SDL_CreateWindow("",
                       SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber),
                       SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber),
                       width, height,
                       (fullscreen ? SDL_WINDOW_FULLSCREEN : 0)
                       | ((screenType == SCREEN_OPENGL) ? SDL_WINDOW_OPENGL : 0) | SDL_WINDOW_SHOWN);
#else
      sdl.window = SDL_CreateWindow("",
                       SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber),
                       SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber),
                       width, height,
                       (fullscreen ? (sdl.desktop.full.display_res ? SDL_WINDOW_FULLSCREEN_DESKTOP : SDL_WINDOW_FULLSCREEN) : 0)
                       | ((screenType == SCREEN_OPENGL) ? SDL_WINDOW_OPENGL : 0) | SDL_WINDOW_SHOWN);
#endif
```
While not perfect, it seems to work. If I set output to OpenGL NB and scaler to Hardware None. Then it does the right thing in full screen mode. Here's a diff against current git; it builds and seems to work (macOS 10.15.6 / MBP 2017).

[retinafix.txt](https://github.com/joncampbell123/dosbox-x/files/5019674/retinafix.txt)

**Comment by DominusExult** [comment]
The macOS SDK 10.15 and newer screwed something up, see https://github.com/libsdl-org/SDL-1.2/issues/839
Something really broke SDL (1.2x and 2) with DOSBox(-x) and the dpi scaling it seems.
It *should* be ok if you compile via SDK 10.14 (or even 10.13) but that is not a solution if you get it via brew and not at all if you are on an Apple M1 mac.

**Comment by Wengier** [comment]
Thanks @DominusExult for pointing out the cause of the issue. So it turns out to be a bug with macOS SDK rather than DOSBox-X.

**Comment by DominusExult** [comment]
Yes, but one that is not likely to be fixed by Apple. Unless someone can create a simple testcase and try to get through via their bugtracker.
But I guess it either needs to be worked around via SDL or DOSBox(-x).
Rather annoying ;(
